File: packages/types/src/settlements.ts
import { z } from 'zod';

/**
 * ABRO_PRD.md §19. Deliberately its own schema, not a splitType branch of
 * createExpenseSchema -- per docs/DECISIONS.md ADR-003, the general
 * expense-create path must never accept splitType: SETTLEMENT. The service
 * builds the underlying Expense + two ExpenseParticipant rows itself; the
 * client only supplies who's being paid, how much, and (optionally) which
 * group the debt belongs to.
 */
export const createSettlementSchema = z.object({
  toUserId: z.string(),
  /** Minor units, as a positive integer string -- see money.ts. */
  amount: z.string().regex(/^[1-9]\d*$/, 'amount must be a positive integer string of minor units'),
  /** Omit for a personal (non-group) settlement. */
  groupId: z.string().optional(),
});
export type CreateSettlementInput = z.infer<typeof createSettlementSchema>;

File: apps/api/internal/apitypes/settlements.go
package apitypes

import (
	"strings"

	"github.com/Substance-k3n/abro/apps/api/internal/httpx"
	"github.com/Substance-k3n/abro/apps/api/internal/money"
)

// CreateSettlementInput mirrors packages/types' createSettlementSchema --
// deliberately its own type, not a splitType branch of CreateExpenseInput,
// per ADR-003.
type CreateSettlementInput struct {
	ToUserID string  `json:"toUserId"`
	Amount   string  `json:"amount"`
	GroupID  *string `json:"groupId"`

	ParsedAmount money.MinorUnits
}

func (in *CreateSettlementInput) Validate() error {
	if strings.TrimSpace(in.ToUserID) == "" {
		return httpx.BadRequest("VALIDATION_ERROR", "toUserId is required")
	}
	parsed, err := money.ParseAmount(in.Amount)
	if err != nil || parsed == 0 {
		return httpx.BadRequest("VALIDATION_ERROR", "amount must be a positive integer string of minor units")
	}
	in.ParsedAmount = parsed
	return nil
}
